File: vbox-app/src/main/java/com/kedzie/vbox/machine/settings/NetworkAdapterFragment.java
package com.kedzie.vbox.machine.settings;

import android.os.Bundle;
import android.os.NetworkOnMainThreadException;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.EditText;
import android.widget.Spinner;

import com.kedzie.vbox.R;
import com.kedzie.vbox.api.IHostNetworkInterface;
import com.kedzie.vbox.api.INetworkAdapter;
import com.kedzie.vbox.api.jaxb.HostNetworkInterfaceType;
import com.kedzie.vbox.api.jaxb.NetworkAdapterPromiscModePolicy;
import com.kedzie.vbox.api.jaxb.NetworkAdapterType;
import com.kedzie.vbox.api.jaxb.NetworkAttachmentType;
import com.kedzie.vbox.app.Tuple;
import com.kedzie.vbox.app.Utils;
import com.kedzie.vbox.task.DialogTask;

import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @apiviz.stereotype fragment
 */
public class NetworkAdapterFragment extends Fragment {
    private static final String TAG = "NetworkAdapterFragment";

	class LoadInfoTask extends DialogTask<INetworkAdapter, Tuple<INetworkAdapter, String[]>> {
		
		public LoadInfoTask() {
			super((AppCompatActivity)getActivity(), _adapter.getAPI(), R.string.progress_loading_data_generic);
		}

		@Override 
		protected Tuple<INetworkAdapter, String[]> work(INetworkAdapter...params) throws Exception {
			INetworkAdapter adapter = params[0];
			params[0].clearCache();
			adapter.getAdapterType(); adapter.getAttachmentType(); adapter.getCableConnected(); 
			adapter.getEnabled(); adapter.getPromiscModePolicy(); adapter.getMACAddress();
			String[] interfaces = null;
			if(adapter.getAttachmentType().equals(NetworkAttachmentType.BRIDGED)) {
				adapter.getBridgedInterface();
				List<IHostNetworkInterface> types = _vmgr.getVBox().getHost().findHostNetworkInterfacesOfType(HostNetworkInterfaceType.BRIDGED);
				interfaces = new String[types.size()+1];
				interfaces[0] = "Not Attached";
				for(int i=0; i<types.size(); i++)
					interfaces[i+1] = types.get(i).getName();
			} else if(adapter.getAttachmentType().equals(NetworkAttachmentType.HOST_ONLY)) {
				adapter.getHostOnlyInterface();
				List<IHostNetworkInterface> types = _vmgr.getVBox().getHost().findHostNetworkInterfacesOfType(HostNetworkInterfaceType.HOST_ONLY);
				interfaces = new String[types.size()+1];
				interfaces[0] = "Not Attached";
				for(int i=0; i<types.size(); i++)
					interfaces[i+1] = types.get(i).getName();
			} else if(adapter.getAttachmentType().equals(NetworkAttachmentType.GENERIC)) {
				adapter.getGenericDriver();
			} else if(adapter.getAttachmentType().equals(NetworkAttachmentType.INTERNAL)) {
				adapter.getInternalNetwork();
			} else if(adapter.getAttachmentType().equals(NetworkAttachmentType.NAT)) {
				interfaces = new String[1];
				interfaces[0] = "Not Attached";
			} 
			return new Tuple<INetworkAdapter, String[]>(adapter, interfaces);
		}
		@Override
		protected void onSuccess(Tuple<INetworkAdapter, String[]> result) {
			super.onSuccess(result);
			_adapter = result.first;
			Log.d(TAG, "Host Interfaces: " + Arrays.toString(result.second));
		    _hostInterfaces = result.second;
		    if(_hostInterfaces!=null) {
		    	_nameAdapter = new ArrayAdapter<String>(getActivity(), android.R.layout.simple_spinner_item, _hostInterfaces);
		    	_nameSpinner.setAdapter(_nameAdapter);
		    }
		    populate();
		}
	}
	
	private INetworkAdapter _adapter;
	private String[] _hostInterfaces;
	
    @BindView(R.id.network_enabled)
	 CheckBox _enabledCheckBox;
    @BindView(R.id.network_name_text)
     EditText _nameText;
    @BindView(R.id.network_mac)
     EditText _macText;
    @BindView(R.id.network_cable_connected)
	 CheckBox _cableConnectedCheckBox;

	@BindView(R.id.network_attached)
	 Spinner _attachmentTypeSpinner;
	private ArrayAdapter<NetworkAttachmentType> _attachmentTypeAdapter;
	private NetworkAttachmentType[] _attachmentTypes = Utils.removeNull(NetworkAttachmentType.values());

	@BindView(R.id.network_name_spinner)
	 Spinner _nameSpinner;
	private ArrayAdapter<String> _nameAdapter;

    @BindView(R.id.network_adapter_type)
	 Spinner _adapterTypeSpinner;
	private ArrayAdapter<NetworkAdapterType> _adapterTypeAdapter;
	private NetworkAdapterType[] _adapterTypes = Utils.removeNull(NetworkAdapterType.values());

	@BindView(R.id.network_promiscuous)
	 Spinner _promiscuousModeSpinner;
	private ArrayAdapter<NetworkAdapterPromiscModePolicy> _promiscuousModeAdapter;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		_adapter = (INetworkAdapter) (savedInstanceState==null ? getArguments().getParcelable(INetworkAdapter.BUNDLE) : savedInstanceState.getParcelable(INetworkAdapter.BUNDLE));
		if(savedInstanceState!=null) {
		    _hostInterfaces = savedInstanceState.getStringArray("interfaces");
		}
	}
	
	@Override
    public void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
        outState.putStringArray("interfaces", _hostInterfaces);
        outState.putParcelable(INetworkAdapter.BUNDLE, _adapter);
    }
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(R.layout.settings_network_adapter, null);
	}

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
        _adapterTypeAdapter = new ArrayAdapter<NetworkAdapterType>(getActivity(), android.R.layout.simple_spinner_item, _adapterTypes);
        _adapterTypeSpinner.setAdapter(_adapterTypeAdapter);
        _attachmentTypeAdapter = new ArrayAdapter<NetworkAttachmentType>(getActivity(), android.R.layout.simple_spinner_item, _attachmentTypes);
        _attachmentTypeSpinner.setAdapter(_attachmentTypeAdapter);
        _promiscuousModeAdapter = new ArrayAdapter<NetworkAdapterPromiscModePolicy>(getActivity(), android.R.layout.simple_spinner_item, NetworkAdapterPromiscModePolicy.values());
        _promiscuousModeSpinner.setAdapter(_promiscuousModeAdapter);
    }

    @Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		if(_adapter.getCache().containsKey("getEnabled"))
			safePopulate();
		else
			new LoadInfoTask().execute(_adapter);
	}

    private void safePopulate() {
        try {
            populate();
        } catch(NetworkOnMainThreadException e) {
            Log.w(TAG, "Populate error. reloading data" + e);
            new LoadInfoTask().execute(_adapter);
        }
    }

	private void populate() {
		boolean enabled = _adapter.getEnabled();
		_enabledCheckBox.setChecked(enabled);
		_enabledCheckBox.setOnCheckedChangeListener(new OnCheckedChangeListener() {
			@Override
			public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
				_adapter.setEnabled(isChecked);
				new LoadInfoTask().execute(_adapter);
			}
		});
		_cableConnectedCheckBox.setEnabled(enabled);
		_macText .setEnabled(enabled);
		_adapterTypeSpinner.setEnabled(enabled);
		_attachmentTypeSpinner.setEnabled(enabled);
		_nameSpinner.setEnabled(enabled);
		_nameText.setEnabled(enabled);
		_promiscuousModeSpinner.setEnabled(enabled);

		_attachmentTypeSpinner.setSelection(Utils.indexOf(_attachmentTypes, _adapter.getAttachmentType()));
		_attachmentTypeSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
				_adapter.setAttachmentType(_attachmentTypeAdapter.getItem(position));
				new LoadInfoTask().execute(_adapter);
			}
			@Override public void onNothingSelected(AdapterView<?> parent) {}
		});
		if(_adapter.getAttachmentType().equals(NetworkAttachmentType.BRIDGED)) {
			_nameSpinner.setVisibility(View.VISIBLE);
			_nameText.setVisibility(View.GONE);
			if(Utils.isEmpty(_adapter.getBridgedInterface()))
				_nameSpinner.setSelection(0);
			else
				_nameSpinner.setSelection(Utils.indexOf(_hostInterfaces, _adapter.getBridgedInterface()));
		} else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.HOST_ONLY)) {
			_nameSpinner.setVisibility(View.VISIBLE);
			_nameText.setVisibility(View.GONE);
			if(Utils.isEmpty(_adapter.getHostOnlyInterface()))
				_nameSpinner.setSelection(0);
			else
				_nameSpinner.setSelection(Utils.indexOf(_hostInterfaces, _adapter.getHostOnlyInterface()));
		} else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.GENERIC)) {
			_nameSpinner.setVisibility(View.GONE);
			_nameText.setVisibility(View.VISIBLE);
			_nameText.setText(_adapter.getGenericDriver());
		} else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.INTERNAL)) {
			_nameSpinner.setVisibility(View.GONE);
			_nameText.setVisibility(View.VISIBLE);
			_nameText.setText(_adapter.getInternalNetwork());
		} else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.NAT)) {
			_nameSpinner.setVisibility(View.VISIBLE);
			_nameText.setVisibility(View.GONE);
			_nameSpinner.setSelection(0);
			_nameSpinner.setEnabled(false);
		} 
		_nameSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
				if(_adapter.getAttachmentType().equals(NetworkAttachmentType.BRIDGED))
					_adapter.setBridgedInterface(_nameAdapter.getItem(position));
				else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.HOST_ONLY))
					_adapter.setHostOnlyInterface(_nameAdapter.getItem(position));
				else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.GENERIC))
					_adapter.setGenericDriver(_nameAdapter.getItem(position));
				else if(_adapter.getAttachmentType().equals(NetworkAttachmentType.INTERNAL))
					_adapter.setInternalNetwork(_nameAdapter.getItem(position));
			}
			@Override public void onNothingSelected(AdapterView<?> parent) {}
		});
		_adapterTypeSpinner.setSelection(Utils.indexOf(_adapterTypes, _adapter.getAdapterType()));
		_adapterTypeSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
				_adapter.setAdapterType(_adapterTypeAdapter.getItem(position));
			}
			@Override public void onNothingSelected(AdapterView<?> parent) {}
		});
		_promiscuousModeSpinner.setSelection(Utils.indexOf(NetworkAdapterPromiscModePolicy.values(), _adapter.getPromiscModePolicy()));
		_promiscuousModeSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
				_adapter.setPromiscModePolicy(_promiscuousModeAdapter.getItem(position));
			}
			@Override public void onNothingSelected(AdapterView<?> parent) {}
		});
		_cableConnectedCheckBox.setChecked( _adapter.getCableConnected());
		_cableConnectedCheckBox.setOnCheckedChangeListener(new OnCheckedChangeListener() {
			@Override
			public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
				_adapter.setCableConnected(isChecked);
			}
		});
		_macText.setText(_adapter.getMACAddress());
		_macText.addTextChangedListener(new TextWatcher() {
			@Override
			public void onTextChanged(CharSequence s, int start, int before, int count) {
				_adapter.setMACAddress(_macText.getText().toString());
			}
			@Override	public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
			@Override public void afterTextChanged(Editable s) {}
		});
	}
}

File: vbox-app/src/main/java/com/kedzie/vbox/machine/settings/NetworkAdapterFragment.kt
package com.kedzie.vbox.machine.settings

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import androidx.fragment.app.Fragment
import com.kedzie.vbox.R
import com.kedzie.vbox.api.IMachine
import com.kedzie.vbox.api.INetworkAdapter
import com.kedzie.vbox.api.jaxb.HostNetworkInterfaceType
import com.kedzie.vbox.api.jaxb.NetworkAdapterPromiscModePolicy
import com.kedzie.vbox.api.jaxb.NetworkAdapterType
import com.kedzie.vbox.api.jaxb.NetworkAttachmentType
import com.kedzie.vbox.app.Utils
import kotlinx.android.synthetic.main.settings_network_adapter.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * @apiviz.stereotype fragment
 */
class NetworkAdapterFragment(arguments: Bundle) : Fragment(), CoroutineScope by CoroutineScope(Dispatchers.Main) {

    init {
        this.arguments = arguments
    }

    private lateinit var machine: IMachine
    private var adapterIndex : Int = 0

    private fun loadInfo() {
        launch {
            val adapterTypes = Utils.removeNull(NetworkAdapterType.values())
            val adapterTypeAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, adapterTypes)
            network_adapter_type.adapter = adapterTypeAdapter
            val attachmentTypes = Utils.removeNull(NetworkAttachmentType.values())
            val attachmentTypeAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, attachmentTypes)
            network_attached.adapter = attachmentTypeAdapter
            val promiscuousModeAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, NetworkAdapterPromiscModePolicy.values())
            network_promiscuous.adapter = promiscuousModeAdapter

            val adapter = machine.getNetworkAdapter(adapterIndex)

            var interfaces = mutableListOf<String>()
            if (adapter.getAttachmentType() == NetworkAttachmentType.BRIDGED) {
                val types = adapter.api.vbox!!.getHost().findHostNetworkInterfacesOfType(HostNetworkInterfaceType.BRIDGED)
                interfaces[0] = "Not Attached"
                for (type in types)
                    interfaces.add(type.getName())
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.HOST_ONLY) {
                val types = adapter.api.vbox!!.getHost().findHostNetworkInterfacesOfType(HostNetworkInterfaceType.HOST_ONLY)
                interfaces.add( "Not Attached")
                for (type in types)
                    interfaces.add(type.getName())
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.NAT) {
                interfaces.add( "Not Attached")
            }

            val nameAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_item, interfaces)
            network_name_spinner.adapter = nameAdapter

            val enabled = adapter.getEnabled()
            network_enabled.isChecked = enabled
            network_enabled.setOnCheckedChangeListener { buttonView, isChecked ->
                launch {
                    adapter.setEnabled(isChecked)
                    loadInfo()
                }
            }
            network_cable_connected.isEnabled = enabled
            network_mac.isEnabled = enabled
            network_adapter_type.isEnabled = enabled
            network_attached.isEnabled = enabled
            network_name_spinner.isEnabled = enabled
            network_name_text.isEnabled = enabled
            network_promiscuous.isEnabled = enabled

            network_attached.setSelection(Utils.indexOf(attachmentTypes, adapter.getAttachmentType()))
            network_attached.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    launch {
                        adapter.setAttachmentType(attachmentTypeAdapter.getItem(position))
                        loadInfo()
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
            if (adapter.getAttachmentType() == NetworkAttachmentType.BRIDGED) {
                network_name_spinner.visibility = View.VISIBLE
                network_name_text.visibility = View.GONE
                if (Utils.isEmpty(adapter.getBridgedInterface()))
                    network_name_spinner.setSelection(0)
                else
                    network_name_spinner.setSelection(interfaces.indexOf(adapter.getBridgedInterface()))
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.HOST_ONLY) {
                network_name_spinner.visibility = View.VISIBLE
                network_name_text.visibility = View.GONE
                if (Utils.isEmpty(adapter.getHostOnlyInterface()))
                    network_name_spinner.setSelection(0)
                else
                    network_name_spinner.setSelection(interfaces.indexOf(adapter.getHostOnlyInterface()))
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.GENERIC) {
                network_name_spinner.visibility = View.GONE
                network_name_text.visibility = View.VISIBLE
                network_name_text.setText(adapter.getGenericDriver())
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.INTERNAL) {
                network_name_spinner.visibility = View.GONE
                network_name_text.visibility = View.VISIBLE
                network_name_text.setText(adapter.getInternalNetwork())
            } else if (adapter.getAttachmentType() == NetworkAttachmentType.NAT) {
                network_name_spinner.visibility = View.VISIBLE
                network_name_text.visibility = View.GONE
                network_name_spinner.setSelection(0)
                network_name_spinner.isEnabled = false
            }
            network_name_spinner.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    launch {
                        if (adapter.getAttachmentType() == NetworkAttachmentType.BRIDGED)
                            adapter.setBridgedInterface(nameAdapter.getItem(position))
                        else if (adapter.getAttachmentType() == NetworkAttachmentType.HOST_ONLY)
                            adapter.setHostOnlyInterface(nameAdapter.getItem(position))
                        else if (adapter.getAttachmentType() == NetworkAttachmentType.GENERIC)
                            adapter.setGenericDriver(nameAdapter.getItem(position))
                        else if (adapter.getAttachmentType() == NetworkAttachmentType.INTERNAL)
                            adapter.setInternalNetwork(nameAdapter.getItem(position))
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
            network_adapter_type.setSelection(Utils.indexOf(adapterTypes, adapter.getAdapterType()))
            network_adapter_type.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    launch { adapter.setAdapterType(adapterTypeAdapter.getItem(position)) }
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
            network_promiscuous.setSelection(Utils.indexOf(NetworkAdapterPromiscModePolicy.values(), adapter.getPromiscModePolicy()))
            network_promiscuous.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    launch { adapter.setPromiscModePolicy(promiscuousModeAdapter.getItem(position)) }
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
            network_cable_connected.isChecked = adapter.getCableConnected()
            network_cable_connected.setOnCheckedChangeListener { buttonView, isChecked -> 
                launch { adapter.setCableConnected(isChecked) } }
            network_mac.setText(adapter.getMACAddress())
            network_mac.addTextChangedListener(object : TextWatcher {
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    launch { adapter.setMACAddress(network_mac.text.toString()) }
                }

                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                override fun afterTextChanged(s: Editable) {}
            })
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.settings_network_adapter, container, false)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        machine = arguments!!.getParcelable(IMachine.BUNDLE)
        adapterIndex = arguments!!.getInt(INetworkAdapter.BUNDLE)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loadInfo()
    }

    override fun onDestroy() {
        cancel()
        super.onDestroy()
    }
}
